src/program/downloaders/alldebrid.py

```diff
@@ -51,594 +34,185 @@ def validate(self) -> bool:
+            return torrent_id
+        raise Exception("Failed to download torrent.")
+
+    def get_cached_containers(
+        self, infohashes: list[str], needed_media: dict, break_pointer: list[bool]
+    ) -> dict:
+        """
+        Get containers that are available in the debrid cache containing `needed_media`
+
+        Parameters:
+        - infohashes: a list of hashes that might contain the data we need
+        - needed_media: a dict of seasons, with lists of episodes, indicating what content is needed
+        - break_pointer: first bool indicates if the needed content was found yet, 2nd pointer indicates if we should break once it's found.
+        """
+        cached_containers = {}
+        response = get_instant_availability(infohashes)
+        magnets = {m.get("hash"): m for m in response}
+
+        for infohash in infohashes:
+            if all(break_pointer):
+                break
+            cached_containers[infohash] = {}
+            magnet = magnets.get(infohash, {})
+            files = magnet.get("files", [])
+            if not files:
                 continue
-    
-            stream = stream_hashes.get(stream_hash_lower)
-            if self._process_providers(item, magnet, stream_hash_lower):
-                return True
-            else:
-                if stream:
-                    stream.blacklisted = True
-    
-        return False
 
-    def _process_providers(self, item: MediaItem, magnet: dict, stream_hash: str) -> bool:
-        """Process providers for an item"""
-        if not magnet or not stream_hash:
-            return False
-    
-        sorted_files = sorted(
-            (file for file in magnet.get("files", [])),
-            key=lambda file: file.get("s", 0),
-            reverse=True
+            # We avoid compressed downloads this way
+            def all_files_valid(files: list) -> bool:
+                filenames = [f.lower() for f, _ in walk_alldebrid_files(files)]
+                return all(
+                    "sample" not in file and file.rsplit(".", 1)[-1] in VIDEO_EXTENSIONS
+                    for file in filenames
+                )
+
+            if all_files_valid(files):
+                # The file_finder needs files to be in a dict, but it doesn't care about the keys
+                container = {
+                    i: dict(filename=name, filesize=size)
+                    for i, (name, size) in enumerate(walk_alldebrid_files(files))
+                }
+                cached_containers[infohash] = self.file_finder.get_cached_container(
+                    needed_media, break_pointer, container
+                )
+                if cached_containers[infohash]:
+                    break_pointer[0] = True
+                    if break_pointer[1]:
+                        break
+
+        return cached_containers
+
+    def get_torrent_names(self, id: str) -> dict:
+        info = get_status(id)
+        return info["filename"], info["filename"]
```

Comment:
AllDebrid doesn't have an "original filename" in the status API. I wasn't sure what it's used for so I just return the filename twice.

Reply:
RD's titles can have alternates, its used when trying trying to symlink so that we can keep track if we cant symlink 1 filename, then we try the other

Reply:
Got it. I can't see anything else that would be useful for that - the `/magnet/upload` endpoint does have a `"filename_original"` key but I haven't seen it populated with anything yet.

If this struggles to symlink I can try adding some kind of cache on the upload endpoint that we can scan for filenames but it's probably better not to persist state in the downloader